### ISSUE TITLE
handle null and non-numeric values gracefully

### ISSUE THREAD

**Issue by VikramTiwari** [metabase/metabase#6615]
- Your databases: BigQuery
- Metabase version: 0.27.2
- Metabase internal database: MySQL (e.x. H2 (default), MySQL, Postgres)

- Create a question where the computation involves a non-numeric value. For example a column with non-zero numerical values divided by a column with zero values. The returned columns are empty which might be fine, but as soon as you try to sort it, it breaks by giving `java.lang.NullPointerException` error.

**Comment by camsaul** [comment]
@VikramTiwari it sounds like this is a problem happing in the database, so we need to modify the query Metabase generates to account for possible `NULL` values. Do you have any suggestions on what the best way to do that would be?

**Comment by camsaul** [comment]
In #7176 (duplicate) using `COALESCE` was suggested. That seems like a reasonable idea

> The way I normally approach `NULL` values is by `COALESCE`.
> 
> So if the query was `SUM(table.numerical_column)` it becomes `SUM(COALESCE(FLOAT(table.numerical_column), 0))`.
> 
> Of course, type inferencing (float/int/string) and then apply proper functions on top will help.
> 
> Also, this works for my use case since most of my values are numerical and representing `NULL` as `0` is fine since they both mean the same thing. But there might be a case where `NULL` is different than just `0`.

**Comment by camsaul** [comment]
Notes on tackling this issue:

This one should be pretty straightforward. We should make sure sorting by columns that contain `NULL` values work for all the DBs we support. For some DBs like BigQuery we will have to automatically `COALESCE` any numeric column when we use it for sorting. Other DBs like H2 and Postgres seem to be able to handle `NULL`s in an `ORDER BY` without any issue. So whether we `COALESCE` for everybody, or just for DBs that break without it, is a good question we'll have to think about.

###### What To Change

For adding `COALESCE()` you can start with [these methods](https://github.com/metabase/metabase/blob/master/src/metabase/driver/sql/query_processor.clj#L689-L699) for `:sql` drivers (either here or overriding them specifically for BigQuery). For MongoDB and others you might not have to do anything if it handles `NULL`s without issue.

###### Tests

 For testing this we have a [`daily-bird-counts` test dataset](https://github.com/metabase/metabase/blob/master/test/metabase/test/data/dataset_definitions.clj#L38) that we can use (definition is [here](https://github.com/metabase/metabase/blob/master/test/metabase/test/data/dataset_definitions/daily-bird-counts.edn)). The test would look something like

```clj
(deftest order-by-nulls-test
  (testing "Check that we can sort by numeric columns that contain NULLs (#6615)"
    (mt/dataset daily-bird-counts
      (mt/test-drivers (mt/normal-drivers)
        (is (= [[2 0 "2018-09-21T00:00:00Z"]
                [3 0 "2018-09-22T00:00:00Z"]
                [9 0 "2018-09-28T00:00:00Z"]]
               (mt/formatted-rows [int int identity]
                 (mt/run-mbql-query bird-count
                   {:order-by [[:asc $count] [:asc $id]]
                    :limit    3}))))))))
```

I would put this test in `metabase.query-processor-test.filter-test`. The actual expected rows may change once the `COALESCE` logic is in place. Depending on the fix we may need to use look at whether the current driver `sorts-nil-first?` -- see [this test](https://github.com/metabase/metabase/blob/aced697af9c7bc71d95427a7a995fd6655bc8ea0/test/metabase/query_processor_test/explicit_joins_test.clj#L87) where we use handle different expected answers from different DBs based on their behavior.  `tx/sorts-nil-first?` is not currently aliased in `mt/` like almost everything else in `tx/`, so feel to add it there if you want.

Ideally we'd want to add a few more tests to make sure we support sorting other column types like Strings that contain NULLs. We have a [`bird-flocks` test dataset](https://github.com/metabase/metabase/blob/master/test/metabase/test/data/dataset_definitions/bird-flocks.edn) that will give you some NULL strings if you join the tables as in that test linked above.

[The only other thing besides numbers and strings we support sorting by is temporal instants](https://github.com/metabase/metabase/blob/master/backend/mbql/src/metabase/mbql/schema.clj#L463) e.g. dates/times/datetimes. AFAIK we don't have any test datasets with `NULL` temporal values but you can define a small one in a similar way to the ones mentioned above and use that.

**Comment by sbelak** [comment]
Since sorting of NULLs is already inconsistent, I'd COALESCE just where necessary. 

One thing to ponder though: perhaps instead of 0, `Infinity` would be a better fill in that way we can be sure NULLs will be at the beginning/end of results rather than potentially somewhere in the middle (imagine sorting by field that goes from -5 to 5 and filling NULLs with 0).

**Comment by noahmoss** [comment]
Currently unable to repro this. I have the `daily-bird-counts` loaded into BigQuery, but I don't see any errors when I try to sort by the `count` column, which contains null values.

**Comment by jeff303** [comment]
As far as sorting goes, the test and related changes added in #17914 seem to confirm that all drivers are able to work with `NULL` values in sorted columns correctly, including BigQuery (legacy and new x.41 driver).

If there are other particular cases of `NULL` values causing errors (as possibly alluded to in the original comment), then let's capture those in detail and address in a new issue.

**Comment by flamber** [comment]
Closing as fixed in 0.41.0, any problems should be created with enough details to reproduce in a new issue.
